var/spack/repos/builtin/packages/mfem/package.py

```diff
@@ -1116,7 +1116,7 @@ def build(self, spec, prefix):
     def check_or_test(self):
         # Running 'make check' or 'make test' may fail if MFEM_MPIEXEC or
         # MFEM_MPIEXEC_NP are not set appropriately.
-        if not self.run_tests:
+        if not self.run_tests and ("+examples" in self.spec):
```

Comment:
This additional check is wrong -- we want to run this quick-check or the full tests (in `else` part) purely based on `self.run_tests` regardless of the `examples` variant.

Reply:
Just to clarify: when I specify `--test=root`, I want to run the full tests which includes building and running all examples, miniapps, and unit tests. The variants `examples` and `miniapps` are meant to specify what is installed, not what's tested.

Reply:
Context: Ran into a linking issue in a build with a Spack built hdf5 on OSX. The MFEM library built/linked properly but this example failed to link. I wanted to try if it failed in the main code by turning off the `examples` variant but noticed it was already off. I continue to battle this linking error.

Reply:
Just another point, other projects allow you to turn off building extra things in their packages as an optimization, eg. examples, exercises, utilities, etc. This is very helpful in speeding up build times and has also been useful in the past to avoid linking issues in upstream libraries when it works downstream customers.

But ultimately, this is your package and if you don't like this change the PR can be closed.

Reply:
@v-dobrev -- could you please clarify something?

My expectation is that `mfem~examples` will not build the mfem examples, but `mfem+examples` will build the examples (and testing and installation are separate issues). Does that match your expectations?

Reply:
> Context: Ran into a linking issue in a build with a Spack built hdf5 on OSX. The MFEM library built/linked properly but this example failed to link. I wanted to try if it failed in the main code by turning off the `examples` variant but noticed it was already off. I continue to battle this linking error.

If the quick-check failed, it just means MFEM was not configured properly and it is missing some flags for correct linking -- that's the purpose of this quick-check (or smoke test).

If you provide some more details (like the link command that failed and the config options passed to mfem's `make config`), we can debug this and fix the issue in the package.

Reply:
> Just another point, other projects allow you to turn off building extra things in their packages as an optimization, eg. examples, exercises, utilities, etc. This is very helpful in speeding up build times and has also been useful in the past to avoid linking issues in upstream libraries when it works downstream customers.
> 
> But ultimately, this is your package and if you don't like this change the PR can be closed.

The quick-check (compiling and linking one simple example) is very quick and it is there to alert users of an issue with the build. The only thing that is not strictly required by MFEM's package and is enabled by default is `zlib` and that can be disabled if needed -- `zlib` is a variant. So by default you get a minimal MFEM build plus `zlib` (which you can turn off if you want).

Reply:
> @v-dobrev -- could you please clarify something?
> 
> My expectation is that `mfem~examples` will not build the mfem examples, but `mfem+examples` will build the examples (and testing and installation are separate issues). Does that match your expectations?

A spec of MFEM with `~examples` (default) will not build and install the examples unless you also enable testing with a Spack flag like `--test=root`/`--test=all`. If you enable testing, the examples will be build and run as part of the testing which is done in the build-tree before installation. After that, if `examples` are enabled, they will be installed (after being optionally re-build, if necessary, e.g. to use the install prefix in a shared build). If `examples` is not enabled, then nothing is (re-)built or installed.

Does this answer your question?

Reply:
Thanks @v-dobrev -- that's helpful.

The only unexpected part is that the testing flags sent to spack can modify whether mfem's examples are built. I'd only expect the test mode run the subset of tests that were previously built.

Reply:
@kennyweiss,

To me it makes sense to do full testing if a user wants to make sure the build of the library is tested. Of course, this is just my point of view. Ultimately, I don't think Chris was running with testing enabled -- what was failing was the quick-check, i.e. the smoke test. To me it makes sense to have that always enabled (it takes 1-2 seconds) unless the full testing is enabled (and that already includes the quick-check).

If you think it will be useful, we can add variant(s) that control what's to be done if testing is disabled (e.g. nothing or smoke test) and what's to be done if testing is enabled (e.g. build and run all tests or just a subset of the tests based on the variants).